### PR TITLE
Filebeat async publisher support

### PR DESCRIPTION
**Pull request by urso**
Add new config option to have filebeat publisher pipeline run in async mode.
This can benefit load-balancing performance, as the publisher pipeline will
be kept more busy with new bulk-events to be published.

### CODE REVIEW

File: filebeat/docs/configuration.asciidoc [hunk 1]
```diff
@@ -307,6 +307,15 @@ filebeat:
 -------------------------------------------------------------------------------------
 
 
+===== publish_async
+
+Experimental!
```
Comment (tsg) [inline filebeat/docs/configuration.asciidoc]: I think there was an asciidoc tag for that? @dedemorton ?

Comment (dedemorton) [inline filebeat/docs/configuration.asciidoc]: I don't know of one, but I'll ask those who are wiser in the ways of asciidoc. @debadair or @palecur Do you know if there is a tag that we could use here for experimental config options? Do we have a convention for documenting experimental options?

@urso By experimental, I assume we mean that customers can try the option, but their mileage may vary (and there's no guarantee that the option will be supported in future versions). If so, maybe we need a sentence that basically conveys "use at your own risk."

File: filebeat/etc/beat.yml [hunk 2]
```diff
@@ -151,6 +151,9 @@ filebeat:
   # Event count spool threshold - forces network flush if exceeded
   #spool_size: 2048
 
+  # Enable async publisher pipeline in filebeat (Experimental!)
```
Comment (ruflin) [inline filebeat/etc/beat.yml]: As it is experimental, I would probably not put it in the config here. Yes I know, somehow breaks our rule so far.

**Comment by ruflin** [comment]
LGTM

File: filebeat/beat/publish.go [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+			p.collect()
+		}
+	}()
+}
+
+// collect collects finished bulk-Events in order and forward processed batches
+// to registrar. Reports to registrar are guaranteed to be in same order
+// as bulk-Events have been received by the spooler
+func (p *asyncLogPublisher) collect() {
```
Comment (ruflin) [inline filebeat/beat/publish.go]: Any idea how we could test if this works as expected?

Comment (urso) [inline filebeat/beat/publish.go]: by having all filebeat system tests use both modes. I think we can easily craft unit test for sync and async mode.

File: filebeat/docs/configuration.asciidoc (hunk 1 above)
Comment (tsg) [inline filebeat/docs/configuration.asciidoc]: See the sources from here: https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-pipeline.html The `experimental[]` string seems to generate a warning. Not sure if that works well in this context, though.

Comment (urso) [inline filebeat/docs/configuration.asciidoc]: >  By experimental, I assume we mean that customers can try the option, but their mileage may vary 

Jup. It's a trade-off between throughput and memory usage when using load-balancing. In future versions I'd consider to remove the flag and enable async mode whenever load-balancing is enabled in output plugins.

**Comment by tsg** [comment]
LGTM

**Comment by ruflin** [comment]
@urso @tsg I would currently suggest to completely remove this feature from the docs and config as it is experimental. We can point people with a specific performance issue specifically to this feature.

**Comment by tsg** [comment]
IMHO, depends on when we release this. If we put it in 1.2, then it should be default off and maybe even undocumented (although I don't particularly like having undocumented options). But if we put it in 5.0-beta1, then we could even make it the default. That's exactly what the betas are for, after all, and we'd lose feedback if we hide the option.

@urso is there a particular reason not to be confident about this?

**Comment by ruflin** [comment]
@tsg Good point. :+1:

**Comment by urso** [comment]
If something goes wrong, we've got a very bad memory leak. But from my manual testings all fine. Will add some unit tests to check the filebeat publisher modes working as expected.

There is no actual reason to have this feature experimental. But users might be unhappy about increased CPU/memory usage if enabled by default. Will need to run some more tests with recent changes in publisher pipeline, as these might have reduced overall memory usage for async mode.

File: filebeat/etc/beat.yml (hunk 2 above)
Comment (urso) [inline filebeat/etc/beat.yml]: Removed the experimental.

**Comment by urso** [comment]
@ruflin @tsg can you review again. Added unit tests for publisher also testing registrar gets data in correct order.

File: libbeat/publisher/client.go [hunk 4]
```diff
@@ -106,13 +115,9 @@ func (c *client) annotateEvent(event common.MapStr) {
...
-func (c *client) getClient(opts []ClientOption) (context, eventPublisher) {
-	var ctx context
-	for _, opt := range opts {
-		ctx = opt(ctx)
-	}
-
-	if ctx.sync {
+func (c *client) getClient(opts []ClientOption) (Context, eventPublisher) {
+	ctx := makeContext(opts)
+	if ctx.Sync {
```
Comment (ruflin) [inline libbeat/publisher/client.go]: This code makes it look like, Async is the default (which I think it isn't). The code is correct, just my first impression.

**Comment by ruflin** [comment]
LGTM

File: libbeat/publisher/client.go (hunk 4 above)
Comment (urso) [inline libbeat/publisher/client.go]: The code you are looking at is in libbeat. Yes in libbeat the default is Async. It is filebeat passing the Sync flag enforing the publisher to do a blocking sync call.

**Comment by cleesmith** [comment]
Just read about this on the [blog](https://www.elastic.co/blog/weekly-beats-generator):
```
This helps with the overall throughput by enabling load balancing between multiple output threads, at the cost of memory usage. With the right settings and enough memory and CPU power, we’ve seen Filebeat pushing around 45K events/s, compared to around 18K before this change.
```
Maybe this isn't the right place to ask, but how would I recreate such a test with the timings mentioned.
Is there a repeatable process in place? I ask, because I would like to do this for unifiedbeat but 
without logstash in between elasticsearch. Of course, IDS alerts are no where near as voluminous
as syslogs, but it would be nice to be as performant as possible with indexing data.
Thanks.

**Comment by urso** [comment]
No fully fledged test framework in place yet. For filebeat testing so far we've got a set of config files with different output options like console, file and logstash output. When using console output we do pipe the output to /dev/null. Logstash itself is configured with beats input only and null output. As input we're using [NASA HTTP logs](http://ita.ee.lbl.gov/html/contrib/NASA-HTTP.html)

Timings are collected as described in this [post](https://discuss.elastic.co/t/insufficient-throughput-from-filebeat/39564/5). Difference is I'm using topbeat and [govarbeat](https://github.com/urso/govarbeat) to collect CPU/memory/throughput stats to elasticsearch/file/console for example.

**Comment by cleesmith** [comment]
Thanks I will give that a try. I guess ```/dev/null``` would indicate the fastest that's possible--if not a bit unrealistic, but still valuable to know.

**Comment by urso** [comment]
yes, using /dev/null to get an idea how fast we can get, as outputs always generate some kind of back presssure.

**Pull request merged**